feat: allow variable interpolation into the custom system prompt

daniel-lxs opened the pull request:
## Context

This PR introduces variable interpolation for custom system prompts loaded from `.roo/system-prompt-[mode]` files. The goal is to allow users to create more dynamic and context-aware custom prompts by injecting values like the current workspace path, mode slug, and language directly into their prompt files using `{{variable_name}}` syntax. This avoids hardcoding potentially changing values within the prompt itself.

For now, only one variable is included in this PR: "workspace" which represents the current working directory, however other variables can be injected to the the custom system prompt with this implementation.

## Implementation

1.  **Defined `PromptVariables` Type:** Added a `PromptVariables` type in `src/core/prompts/sections/custom-system-prompt.ts` to explicitly define the structure and potential keys (`workspace` initially) for variables that can be interpolated.
2.  **Created `interpolatePromptContent` Function:** Implemented a helper function within `custom-system-prompt.ts` that takes the raw prompt content and a `PromptVariables` object. It uses a regex (`/\{\{key\}\}/g`) to find and replace placeholders with their corresponding values from the object. Undefined variables in the object or placeholders not matching a key are left untouched in the content.
3.  **Modified `loadSystemPromptFile`:** Updated the signature of `loadSystemPromptFile` in `custom-system-prompt.ts` to accept the `PromptVariables` object. It now reads the raw file content using `safeReadFile` and then passes the content and variables to `interpolatePromptContent` before returning the result.
4.  **Updated Caller (`SYSTEM_PROMPT`):** Refactored the main `SYSTEM_PROMPT` function in `src/core/prompts/system.ts`. It now gathers the necessary context values into a `variablesForPrompt` object (typed as `PromptVariables`) and passes this object when calling the updated `loadSystemPromptFile`.
5.  **Added Tests:** Created a new Jest test suite `src/core/prompts/sections/__tests__/custom-system-prompt.test.ts`. This suite mocks `fs/promises.readFile` and includes tests for various scenarios, including successful interpolation, handling of missing variables/placeholders, multiple occurrences, and file system errors (`ENOENT`, empty files), ensuring the feature's robustness.

**Refactoring Notes:** The implementation approach was iterated upon. Initial ideas involved performing interpolation within `system.ts` and using separate constant files, but the final approach centralizes the core interpolation logic within `custom-system-prompt.ts` for better module cohesion, as preferred.

## Screenshots

| before                                      | after                                       |
| :------------------------------------------ | :------------------------------------------ |
| N/A (Logic change, no visual UI difference) | N/A (Logic change, no visual UI difference) |

## How to Test

1.  **Create/Modify a Custom Prompt:** In your workspace, create or edit a file named `.roo/system-prompt-code` (or for any other mode slug).
2.  **Add Placeholders:** Add the following lines to the file:
    ```
    Current Workspace: {{workspace}}
    Undefined Variable Test: {{my_undefined_var}}
    ```
3.  **Run Roo:** Start a new Roo chat session using the 'Code' mode (or whichever mode corresponds to your file).
4.  **Verify Interpolation:**
    *   Observe the system prompt being used (this might require adding temporary logging in `SYSTEM_PROMPT` before it's sent to the model, or using debugging tools).
    *   Confirm that `{{workspace}}` has been replaced with the actual absolute path to your workspace.
    *   Confirm that `{{my_undefined_var}}` remains unchanged in the prompt.
5.  **Test File Not Found:** Temporarily rename or delete the `.roo/system-prompt-code` file and start a new chat in 'Code' mode. Verify that Roo falls back to the default generated system prompt without errors.

<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Introduces variable interpolation in custom system prompts, allowing dynamic content based on the workspace path, with updates to `loadSystemPromptFile` and new tests in `custom-system-prompt.test.ts`.
> 
>   - **Behavior**:
>     - Adds variable interpolation to custom system prompts using `{{variable_name}}` syntax in `custom-system-prompt.ts`.
>     - Supports `workspace` variable for current working directory.
>     - Undefined variables remain unchanged in the prompt.
>   - **Functions**:
>     - `interpolatePromptContent`: Replaces placeholders with values from `PromptVariables`.
>     - `loadSystemPromptFile`: Updated to accept `PromptVariables` and return interpolated content.
>     - `SYSTEM_PROMPT`: Refactored to pass `workspace` as a variable for interpolation.
>   - **Tests**:
>     - New test suite `custom-system-prompt.test.ts` for `loadSystemPromptFile`.
>     - Tests include file not found, empty content, successful interpolation, and handling of undefined variables.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=RooVetGit%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for 1875137c99418f1662e78145b9e39d6b87ef0665. You can [customize](https://app.ellipsis.dev/RooVetGit/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

mrubens reviewed (approved):
Nice! Makes sense to me.

NamesMT commented:
Hi @mrubens, @daniel-lxs, FYI there is a PR with similar feature: #2366
I think it would be great if we port over some variables included in #2366 like:
```
"{{ROLE_DEFINITION}}"
"{{SHARED_TOOL_USE_SECTION}}"
"{{TOOL_DESCRIPTIONS}}"
"{{TOOL_USE_GUIDELINES}}"
"{{MCP_SERVERS_SECTION}}"
"{{CAPABILITIES_SECTION}}"
"{{MODES_SECTION}}"
```

This would allows us to craft foot gun modes easier and more compatible between Roo versions, as well as allowing the modes to automatically inherit MCP and available tools without manually specifying them.

daniel-lxs commented:
> Hi @mrubens, @daniel-lxs, FYI there is a PR with similar feature: #2366
> I think it would be great if we port over some variables included in #2366 like:
> ```
> "{{ROLE_DEFINITION}}"
> "{{SHARED_TOOL_USE_SECTION}}"
> "{{TOOL_DESCRIPTIONS}}"
> "{{TOOL_USE_GUIDELINES}}"
> "{{MCP_SERVERS_SECTION}}"
> "{{CAPABILITIES_SECTION}}"
> "{{MODES_SECTION}}"
> ```
> 
> This would allows us to craft foot gun modes easier and more compatible between Roo versions, as well as allowing the modes to automatically inherit MCP and available tools without manually specifying them.

I saw the possibility of adding these but I figured out it would break every single footgun prompt for every user if we removed the role definition and custom instructions from the custom system prompt.

So I added only the variables that are not currently accessible to users.

NamesMT commented:
> I saw the possibility of adding these but I figured out it would break every single footgun prompt for every user if we removed the role definition and custom instructions from the custom system prompt.
> 
> So I added only the variables that are not currently accessible to users and that are not huge sections of the default system prompt like tooling and MCP servers.

Ah right, my bad, I just re-checked the docs and looks like `roleDefinition` `customInstructions` is already always included with footgun by default, in that case, we can skip em and add the others like mcp and tool, capabilities, which would be amazing.
